inference-engine/src/mkldnn_plugin/emitters/jit_snippets_emitters.hpp

```diff
@@ -0,0 +1,573 @@
+};
+
+class NopEmitter : public jit_emitter {
+public:
+    NopEmitter(mkldnn::impl::cpu::x64::jit_generator* h, mkldnn::impl::cpu::x64::cpu_isa_t isa, const std::shared_ptr<ov::Node>& n)
+    : jit_emitter(h, isa, n) {
+    }
+
+    size_t get_inputs_num() const override {return 0;}
+
+private:
+    void emit_impl(const std::vector<size_t>& in,
+                   const std::vector<size_t>& out,
+                   const std::vector<size_t>& pool,
+                   const std::vector<size_t>& gpr,
+                   const MKLDNNPlugin::emitter_context *emit_context) const override {
+    }
+};
+
+class FakeBroadcastEmitter : public jit_emitter {
+public:
+    FakeBroadcastEmitter(mkldnn::impl::cpu::x64::jit_generator* h, mkldnn::impl::cpu::x64::cpu_isa_t isa, const std::shared_ptr<ov::Node>& n)
+    : jit_emitter(h, isa, n) {
+        if (n->get_input_shape(0).empty())
+            use_broadcast = true;
+        else if (*n->get_input_shape(0).rbegin() != *n->get_output_shape(0).rbegin())
+            use_broadcast = true;
+        else
+            use_broadcast = false;
+    }
+    size_t get_inputs_num() const override {return 1;}
+
+private:
+    void emit_impl(const std::vector<size_t>& in,
+              const std::vector<size_t>& out,
+              const std::vector<size_t>& pool,
+              const std::vector<size_t>& gpr,
+              const MKLDNNPlugin::emitter_context *emit_context) const override {
+        if (host_isa_ == dnnl::impl::cpu::x64::sse41) {
+            emit_isa<dnnl::impl::cpu::x64::sse41>(in, out);
+        } else if (host_isa_ == dnnl::impl::cpu::x64::avx2) {
+            emit_isa<dnnl::impl::cpu::x64::avx2>(in, out);
+        } else if (host_isa_ == dnnl::impl::cpu::x64::avx512_common) {
+            emit_isa<dnnl::impl::cpu::x64::avx512_common>(in, out);
+        } else {
+            IE_THROW() << host_isa_;
+            assert(!"unsupported isa");
+        }
+    }
+
+    template <dnnl::impl::cpu::x64::cpu_isa_t isa>
+    void emit_isa(const std::vector<size_t> &in, const std::vector<size_t> &out) const {
+        using Vmm = typename dnnl::impl::utils::conditional3<isa == dnnl::impl::cpu::x64::sse41,
+                                    Xmm, isa == dnnl::impl::cpu::x64::avx2, Ymm, Zmm>::type;
+        Vmm vmm_src0 = Vmm(in[0]);
+        Vmm vmm_dst  = Vmm(out[0]);
+
+        if (use_broadcast) {
+            h->uni_vbroadcastss(vmm_dst, Xmm(in[0]));
+        } else {
+            h->uni_vmovups(vmm_dst, vmm_src0);
+        }
+    }
+
+private:
+    bool use_broadcast;
+};
+
+class ScalarEmitter : public jit_emitter {
+public:
+    ScalarEmitter(mkldnn::impl::cpu::x64::jit_generator* h, mkldnn::impl::cpu::x64::cpu_isa_t isa, const std::shared_ptr<ov::Node>& n)
+    : jit_emitter(h, isa, n) {
+        auto out_shape = n->output(0).get_tensor().get_shape();
+        if (out_shape == ov::Shape() || ov::shape_size(out_shape) == 1) {
```

Comment:
Why we need this condition? Is it possible to have scalar for other shapes? If no seems we need exception here.

Reply:
I added the exception, as well as check that the shape is static. However, do you think it is a proper place for these checks? It is probably better to move them inside snippets::op::Scalar constructor, since they are not plugin-specific

Reply:
Agree